File: nim/bin2file/bin2file.nim
#
# Put binary portion from file to file
# Copyright (c) 2021 Dmitry Stefankov
#
import os
import strutils
#import parseutils
import streams

echo "program name: ", getAppFilename()
#echo "Arguments:"
#for arg in commandLineParams():
#  echo arg
#echo paramCount(), " ", paramStr(1)
let args_count =  paramCount()
if args_count < 5:
  echo "Put binary portion from file, version 1.0"
  echo "Copyright (C) 2021 Dmitry Stefankov. All Rights Reserved."
  echo "Usage: bin2file.exe infile outfile in_offset in_size out_offset"
  quit(1)

let infile = paramStr(1)
let outfile = paramStr(2)
echo "infile: ", infile
echo "outfile: ", outfile
let in_offset = paramStr(3).parseInt()
let in_length = paramStr(4).parseInt()
let out_offset = paramStr(5).parseInt()
echo "in_offset: ", in_offset
echo "in_length: ", in_length
echo "out_offset: ", out_offset

let in_filesize = getFileSize infile
echo "in_filesize: ", in_filesize

let out_filesize = getFileSize outfile
echo "out_filesize: ", out_filesize

var mybuf = alloc(in_length)
var fs = newFileStream(infile, fmRead)
fs.setPosition(in_offset)
var readlen = readData(fs, mybuf, in_length)
echo "readlen: ",readlen
fs.close()

var newfs = newFileStream(outfile, fmReadWriteExisting)
newfs.setPosition(out_offset)
writeData(newfs, mybuf, in_length)
newfs.close()

File: R/bin2file/bin2file.r
# Rscript[.exe] bin2file.r test.in test.out 35 67 13

Sys.setenv(LANG='C')

main <- function() {
  args <- commandArgs(trailingOnly = TRUE)
  in_filename <- args[1]
  out_filename <- args[2]
  in_offset <- args[3]
  in_size <- args[4]
  out_offset <- args[5]
  print(in_filename)
  print(out_filename)
  print(in_offset)
  print(in_size)
  print(out_offset)
  cx <- file(in_filename, "rb")
  seek(cx, in_offset)
  d <- readBin(cx,"raw",in_size,size=1)
  close(cx)
  con <- file(out_filename, "r+b")
  seek(con, out_offset, origin = "start", rw = "write")
  writeBin(d, con, size = 1)
  close(con)
}

if (length(commandArgs(trailingOnly = TRUE)) == 0) {
  print("Put binary portion to file, version 1.0")
  print("Copyright (C) 2024 Dmitry Stefankov. All Rights Reserved.");
  print("Usage: bin2file infile outfile in_offset in_size out_offset");
  print("infile     - input filename");
  print("outfile    - output filename");
  print("in_offset  - byte offset on input stream (default=0)");
  print("in_size    - copy N bytes from input stream (default=0)");
  print("out_offset - byte offset on output stream (default=0)")
  quit(status=1)
}

main()

